### DISCUSSION TITLE
How to correctly handle null values passed as method argumets?

### DISCUSSION THREAD

**Discussion by uniqx** [jni-rs/jni-rs#452] [Q&A]
Hi, I need to process a JString value passed as method argument, which may or may not be null. Here's a rough outline of how I imagine my code might look like. (Although the `unwrap` call crashes when the passed value is null.)

```
#[no_mangle]
#[allow(non_snake_case)]
pub extern "system" fn Java_com_example_MyClass_myMethod<'local>(
    mut env: JNIEnv<'local>,
    _class: JClass<'local>,
    myValue: JString<'local>,
) -> jstring {

    let myValue: String = env
        .get_string(&myValue)
        .unwrap()
        .to_string_lossy()
        .into_owned();

    // if myvalue is null
    handleValue(None);
    // else
    handleValue(Some(myValue));
}
```

I couldn't find any documentation on how jni-rs suggests to deal handling null values.

**Comment by rib** [comment]
You can use [`JObject::null()`](https://docs.rs/jni/latest/jni/objects/struct.JObject.html#method.null) for a null object reference.

There has been discussion in the past around whether to use `Option::None` for `null` but considering how common it is to pass null object references in Java and considering needing to map from the `sys` types to `jni` references I think it would be more awkward (or at least a lot more verbose) to have to wrap all object references in an `Option`.

Here's one of the older discussions I was thinking of: https://github.com/jni-rs/jni-rs/issues/148

**Comment by uniqx** [comment]
I ended up unwrapping the JString variable into and Option<String>:

```
    let myValue: Option<String> = match env.get_string(&myValue) {
        Ok(v) => Some(v.to_string_lossy().into_owned()),
        Err(_) => None,
    };
```

So when my java method gets called with the String parameter set to null: `(new MyClass()).myMethod(null)` it'll convert to Option<String>::None on the rust side.

**Reply by argv-minus-one** [reply]
That will yield `None` if *any* error occurs in getting the string's value. That may or may not be what you really want. If you only want to have `None` if the Java `String` is null, then like @rib said, call `is_null` on the `JString`. Perhaps something like this:

```rust
if myValue.is_null() {
    let myValue: String = env
        .get_string(&myValue)
        .unwrap()
        .to_string_lossy()
        .into_owned();
    handleValue(Some(myValue));
}
else {
    handleValue(None);
}
```

**Comment by rib** [comment]
As @argv-minus-one replied above; the intended way to check for a `null` reference it to use `.is_null()` and not to automatically map nulls into errors or `Option::None`.

In the past there had been some inconsistency in the API, with some parts returning `Option`s and other APIs returning errors for null that didn't represent a logical failure.

At this point the API is (hopefully) fairly consistent with not treating Java null values as anything special (any Java object reference may be null and that often doesn't imply an error).

Any `Reference` type (such as `JString`) can be initialised as null references like `JString::null()` or via `Default::default()` and can be tested for null via `Reference::is_null()` (trait method), something like:

```rust
use jni::{jni_mangle, objects::{JClass, JString}, EnvUnowned};
use jni::refs::Reference as _;

#[jni_mangle("com.example.MyClass")]
pub fn my_method<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    my_value: JString<'local>,
) -> JString<'local> {
    env.with_env(|env| -> jni::errors::Result<()> {
        let opt_value = if my_value.is_null() { Some(my_value) } else { None };
        Ok(handle_value(opt_value))
    }).resolve::<ThrowRuntimeExAndDefault>()
}
```
